core/src/main/java/io/grpc/NameResolverRegistry.java

```diff
@@ -0,0 +1,183 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Registry of {@link NameResolverProvider}s.  The {@link #getDefaultRegistry default instance}
+ * loads providers at runtime through the Java service provider mechanism.
+ *
+ * @since 1.21.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4159")
+@ThreadSafe
+public final class NameResolverRegistry {
+  private static final Logger logger = Logger.getLogger(NameResolverRegistry.class.getName());
+  private static NameResolverRegistry instance;
+
+  private final NameResolver.Factory factory = new NameResolverFactory();
+
+  @GuardedBy("this")
+  private final LinkedHashSet<NameResolverProvider> allProviders = new LinkedHashSet<>();
+  /** Immutable, sorted version of {@code allProviders}. Is replaced instead of mutating. */
+  @GuardedBy("this")
+  private List<NameResolverProvider> effectiveProviders = Collections.emptyList();
+
+  /**
+   * Register a provider.
+   *
+   * <p>If the provider's {@link NameResolverProvider#isAvailable isAvailable()} returns
+   * {@code false}, this method will throw {@link IllegalArgumentException}.
+   *
+   * <p>Providers will be used in priority order. In case of ties, providers are used in
+   * registration order.
+   */
+  public synchronized void register(NameResolverProvider provider) {
+    addProvider(provider);
+    refreshProviders();
+  }
+
+  private synchronized void addProvider(NameResolverProvider provider) {
+    checkArgument(provider.isAvailable(), "isAvailable() returned false");
+    allProviders.add(provider);
+  }
+
+  /**
+   * Deregisters a provider.  No-op if the provider is not in the registry.
+   *
+   * @param provider the provider that was added to the register via {@link #register}.
+   */
+  public synchronized void deregister(NameResolverProvider provider) {
+    allProviders.remove(provider);
+    refreshProviders();
+  }
+
+  private synchronized void refreshProviders() {
+    List<NameResolverProvider> providers = new ArrayList<NameResolverProvider>(allProviders);
+    // sort() must be stable, as we prefer first-registered providers
+    Collections.sort(providers, new Comparator<NameResolverProvider>() {
+      @Override
+      public int compare(NameResolverProvider p1, NameResolverProvider p2) {
+        return p1.priority() - p2.priority();
```

Comment:
I was under the impression that higher priority should be used first, and this is how `LoadBalancerRegistry` does.

Reply:
I was originally going to use the comparator in ServiceProviders. As things changed around I ended up missing a Collections.reverseOrder. I had forgotten higher preferred, so didn't notice in the tests. Fixed.